File: scripts/process_fcs.R
#!/usr/bin/env Rscript

library(optparse)
library(flowCore)
library(ggplot2)
library(uwot)

option_list <- list(
  make_option(c("-i", "--input"), type="character", help="Input FCS file"),
  make_option(c("-o", "--output"), type="character", help="Output FCS file"),
  make_option(c("-p", "--plot"), type="character", help="Output plot file"),
  make_option(c("-c", "--channels"), type="character", default=NULL,
              help="Optional channels txt file specifying which channels to use")
)

opt <- parse_args(OptionParser(option_list=option_list))

if (is.null(opt$input) || is.null(opt$output) || is.null(opt$plot)) {
  stop("Input, output and plot must be provided")
}

ff <- read.FCS(opt$input, transformation = FALSE)
params <- parameters(ff)
desc <- pData(params)$desc

if(!is.null(opt$channels)){
  ch <- read.table(opt$channels, header=TRUE, sep="\t", stringsAsFactors = FALSE)
  selected <- ch$desc[ch$use == 1]
  present <- selected[selected %in% desc]
  if(length(present) == 0){
    channels <- which(!is.na(desc) & !grepl("^FSC|^SSC", desc, ignore.case = TRUE))
  }else{
    print("some channels same")
    channels <- which(desc %in% selected)
  }
}else{
  channels <- which(!is.na(desc) & !grepl("^FSC|^SSC", desc, ignore.case = TRUE))
}
expr <- exprs(ff)[, channels, drop=FALSE]
expr_trans <- asinh(expr / 5)

umap_res <- umap(expr_trans, n_components = 2)
km <- kmeans(umap_res, centers = 5)

new_expr <- cbind(exprs(ff), UMAP1=umap_res[,1], UMAP2=umap_res[,2], Cluster=km$cluster)
new_ff <- flowFrame(new_expr)

write.FCS(new_ff, filename = opt$output)

df <- data.frame(UMAP1=umap_res[,1], UMAP2=umap_res[,2], Cluster=factor(km$cluster))
plt <- ggplot(df, aes(x=UMAP1, y=UMAP2, color=Cluster)) + geom_point(size=0.5) + theme_minimal()

ggsave(opt$plot, plot=plt)

File: scripts/process_fcs.py
#!/usr/bin/env python

import argparse
import numpy as np
import pandas as pd
import umap
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt
from FlowCal.io import FCSData
from fcswrite import write_fcs

def main():
    parser = argparse.ArgumentParser(description="Process FCS file")
    parser.add_argument("-i", "--input", required=True, help="Input FCS file")
    parser.add_argument("-o", "--output", required=True, help="Output FCS file")
    parser.add_argument("-p", "--plot", required=True, help="Output plot file")
    parser.add_argument("-c", "--channels", default=None, help="Optional channels.txt")
    args = parser.parse_args()

    ff = FCSData(args.input)
    expr = np.array(ff, dtype=float)

    if args.channels:
        channels_df = pd.read_csv(args.channels, sep="\t")
        selected = channels_df.loc[channels_df["use"] == 1, "desc"].tolist()
        available_channels = ff.channel_labels()
        present = [ch for ch in selected if ch in available_channels]
        indices = [available_channels.index(ch) for ch in selected]
    else:
        available_channels = ff.channel_labels()
        indices = [i for i, desc in enumerate(available_channels) if
                   desc and not any(s in desc.upper() for s in ["FSC", "SSC"])]

    X = expr[:, indices]
    X_asinh = np.arcsinh(X / 5)

    umap_res = umap.UMAP(n_components=2).fit_transform(X_asinh)

    km = KMeans(n_clusters=5, random_state=42).fit(umap_res)

    new_data = np.hstack([expr, umap_res, km.labels_[:, None]])

    new_channels = ff.channel_labels() + ["UMAP1", "UMAP2", "Cluster"]
    write_fcs(
        filename=args.output,
        chn_names=new_channels,
        data=new_data
    )

    plt.figure(figsize=(6,6))
    plt.scatter(umap_res[:,0], umap_res[:,1], c=km.labels_, cmap='tab10', s=1)
    plt.xlabel("UMAP1")
    plt.ylabel("UMAP2")
    plt.title("UMAP colored by cluster")
    plt.tight_layout()
    plt.savefig(args.plot, dpi=150)

if __name__ == "__main__":
    main()